### PR TITLE
Map::bash rewrite, fixes "flying ceiling"

### PR DESCRIPTION
**Pull request by Coolthulhu**
I wanted mostly to fix the "flying ceiling" bug I introduced with floor smashing, but I noticed that `map::bash` is quite low quality, so I rewrote it.

I fixed the bug where "destroy" type effects (collapsing buildings) don't destroy ceiling if they can't also destroy the floor below it.

I rewrote the bashing functions as follows:
- Split the bashing code into separate functions: `bash_field` (web), `bash_items` (wrecks glass), `bash_vehicle` and `bash_ter_furn`. They are passed a reference to structure that contains parameters to bashing and results of it (rather than having them all as separate parameters and returning a `std::pair<bool, bool>).
- Inverted `if`s where possible, a great bulk of the bashing code was in giant if-else braces.
- Changed tent code from a hardcoded+copypasted code to map bashing data. This will also allow mods to add own tent types. By "tent" I mean a structure that brings down several tiles of selected types when it is bashed down, but without crashing down or recursively bashing.
- Changed parameter order in `map::bash` (moved `bashing_vehicle` to the end, because it is used only once in the code)

### CODE REVIEW

File: src/map.cpp [hunk 1]
```diff
@@ -3017,315 +3017,300 @@ ter_id map::get_roof( const tripoint &p, const bool allow_air )
...
+            sound_volume = std::min(int(smin * 1.5), smax);
+        } else {
+            sound_volume = sound_vol;
+        }
+    }
+
+    sound = _(bash->sound.c_str());
+    // Set this now in case the ter_set below changes this
+    const bool collapses = has_flag("COLLAPSES", p) && smash_ter;
+    const bool supports = has_flag("SUPPORTS_ROOF", p) && smash_ter;
```
Comment (BevapDin) [inline src/map.cpp]: If you want better performance, the boolean variable should be the first and the function the second operand. If the variable already contains `false`, the function won't be called at all.

File: src/map.cpp [hunk 2]
```diff
@@ -3017,315 +3017,300 @@ ter_id map::get_roof( const tripoint &p, const bool allow_air )
...
+        furn_id center_type = f_null;
+        
+        // Find the center of the tent
+        // First check if we're not currently bashing the center
+        if( centers.count( furn( p ) ) > 0 ) {
+            tentp = p;
+            center_type = furn( p );
+        } else {
+            for( const tripoint &pt : points_in_radius( p, bash->collapse_radius ) ) {
+                const furn_id &f_at = furn( pt );
```
Comment (BevapDin) [inline src/map.cpp]: Nothing wrong here, only nitpicking: A `furn_id` is just an int, storing a reference to it is not really faster than storing the id itself. Also `map::furn` returns a value and not a reference anyway.

File: src/map.cpp [hunk 3]
```diff
@@ -3017,315 +3017,300 @@ ter_id map::get_roof( const tripoint &p, const bool allow_air )
...
-                } else {
-                    tripoint below( p.x, p.y, p.z - 1 );
-                    const auto roof = get_roof( below, bash_floor && ter_at( below ).movecost != 0 );
-                    ter_set( p, roof );
+        tripoint tentp = tripoint_min;
+        furn_id center_type = f_null;
+        
+        // Find the center of the tent
+        // First check if we're not currently bashing the center
+        if( centers.count( furn( p ) ) > 0 ) {
```
Comment (BevapDin) [inline src/map.cpp]: Why is this case checked separately? Isn't the center point included in the points returned by `points_in_radius`? I think the scenario where someone bashes the center of the tent is not much more likely then bashing the outsides of it. Checking this separately seems (in my opinion) not useful.

File: src/map.cpp [hunk 4]
```diff
@@ -3017,315 +3017,300 @@ ter_id map::get_roof( const tripoint &p, const bool allow_air )
...
+        str, silent, destroy, bash_floor, (float)rng_float( 0, 1.0f ), false, false, false
+    };
+
+    // Remove webs
+    bash_field( p, bsh );
+
+    // Bash glass items
+    bash_items( p, bsh );
+
+    // Bash vehicle
```
Comment (BevapDin) [inline src/map.cpp]: Sorry for the rant.

The function is named `bash_vehicle` and the comment says "bash vehicle" - what a surprising and useful comment. I had guessed that `bash_vehicle` is used to format the hard drive. Good thing the comment informed me what it actual does. /irony

Similar the comments above contain things about the implementation of the called function. `bash_field` does remove webs, but what if someone changes it to also remove other fields? Now the comment here is at least incomplete. Similar with `bash_items`, what if someone adds code to damage other items as well? (Ceramic may be fragile, plastic can also be bashed.)

In short: I think the function names speak for themselves and don't need to be commented here.

**Comment by BevapDin** [comment]
The comments above are all nitpicking. It still looks fine.

File: src/map.cpp (hunk 3 above)
Comment (Coolthulhu) [inline src/map.cpp]: It changes the order of checked tiles to be a bit more predictable. With separate case, bashing a 3x3 block of tent center tiles will destroy the bashed tile and 3x3 area surrounding it, but in a generic case it would select the 2x2 sub-block in the upper left and bash that.

Comment (BevapDin) [inline src/map.cpp]: A 3x3 block of tent center tiles is already invalid terrain, how do you get that? And what happens to the tent wall tiles around it (if there are any)?

Comment (Coolthulhu) [inline src/map.cpp]: FEMA camp big tents can be "cut" into smaller pieces. While a 3x3 block may not be possible, 3x3 with one tile removed almost certainly is.

Wall tiles would be removed around the chosen center.
